px-map-behavior-layer-IMS.es6.js

```diff
@@ -191,13 +191,376 @@
+          let covered = true;
+          for (let split of splits) {
+            if (!this._boundsCovered(split, newBoundsArray)) {
+              covered = false;
+              break;
+            }
+          }
+          return covered;
+        }
+      }
+      return false;
+    },
+
+    /**
+     * Returns the current map bounds.
+     */
+    _getMapBounds() {
+      const mapBounds = this.parentNode.elementInst.getBounds();
+      return {
+        minX: mapBounds._southWest.lng,
+        minY: mapBounds._southWest.lat,
+        maxX: mapBounds._northEast.lng,
+        maxY: mapBounds._northEast.lat
+      };
+    },
+
+    /**
+     * Returns an array of position pairs from the feature coordinates.
+     */
+    _getFeatureCoords(feature) {
+      let coords = [];
+      const featureCoords = feature.geometry.coordinates;
+      switch (feature.geometry.type) {
+        case 'Point':
+          coords.push(featureCoords);
+          break;
+        case 'LineString':
+        case 'MultiPoint':
+          coords = featureCoords;
+          break;
+        case 'Polygon':
+        case 'MultiLineString':
+          featureCoords.forEach(someCoords => {
+            coords.push.apply(coords, someCoords);
+          });
+          break;
+        case 'MultiPolygon':
+          featureCoords.forEach(someCoords => {
+            someCoords.forEach(someMoreCoords => {
+              coords.push.apply(coords, someMoreCoords);
+            });
+          });
+      }
+      return coords;
+    },
+
+    /**
+     * Maintains the cache of features.
+     * Removes older bounds from the cache if exceeding max cache size.
+     * Adds new bounds and features to the cache.
+     */
+    _updateFeatureCache() {
+      if (!this.enableCache) {
+        return;
+      }
+
+      const mapBounds = this._getMapBounds();
+
+      /*
+       * Remove older bounds if exceeding max cache size - but only if they don't
+       * overlap with the current bounds.
+       */
+      const n = this.maxBoundsCacheSize - 2;
+      for (let i = this._boundsCache.length - 1; i > n; i--) {
+        const bounds = this._boundsCache[i];
+        if (!this._boundsOverlaps(bounds, mapBounds)) {
+          const elementsToRemove = this._featureTreeCache.search(bounds);
+          elementsToRemove.forEach(element => {
+            let remove = true;
+            for (let b of this._boundsCache) {
+              if (b !== bounds && this._boundsInteracts(element, b)) {
+                remove = false;
+                break;
+              }
+            }
+            if (remove) {
+              this._featureCache.delete(element.feature.id);
+              this._featureTreeCache.remove(element);
+            }
+          });
+          this._boundsCache.splice(i, 1);
+        }
+      }
+
+      // Add new features from the current feature collection to the cache
+      const features = this.featureCollection.features;
```

Comment:
Can this block be moved to its own function? It would help reduce the size of this one.